plugin-server/src/worker/ingestion/event-pipeline/runner.ts

```diff
@@ -80,116 +50,165 @@ export class EventPipelineRunner {
+            })
+            return result
+        } catch (error) {
+            if (error instanceof DependencyUnavailableError) {
+                // If this is an error with a dependency that we control, we want to
+                // ensure that the caller knows that the event was not processed,
+                // for a reason that we control and that is transient.
+                throw error
+            }
+
+            return { lastStep: error.step, args: [], error: error.message }
+        }
     }
 
-    private async runPipeline<Step extends StepType, ArgsType extends StepParameters<EventPipelineStepsType[Step]>>(
-        name: Step,
-        ...args: ArgsType
+    async runBufferEventPipelineSteps(
+        event: PluginEvent,
+        personContainer: LazyPersonContainer
     ): Promise<EventPipelineResult> {
-        let currentStepName: StepType = name
-        let currentArgs: any = args
+        const processedEvent = await this.runStep(pluginsProcessEventStep, [this, event])
 
-        while (true) {
-            const timer = new Date()
-            try {
-                const stepResult = await this.runStep(currentStepName, ...currentArgs)
-
-                this.hub.statsd?.increment('kafka_queue.event_pipeline.step', { step: currentStepName })
-                this.hub.statsd?.timing('kafka_queue.event_pipeline.step.timing', timer, { step: currentStepName })
-
-                if (stepResult) {
-                    ;[currentStepName, currentArgs] = stepResult
-                } else {
-                    this.hub.statsd?.increment('kafka_queue.event_pipeline.step.last', {
-                        step: currentStepName,
-                        team_id: String(this.originalEvent?.team_id),
-                    })
-                    return {
-                        lastStep: currentStepName,
-                        args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    }
-                }
-            } catch (error) {
-                await this.handleError(error, currentStepName, currentArgs)
-                return {
-                    lastStep: currentStepName,
-                    args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    error: error.message,
-                }
+        if (processedEvent != null) {
+            const [normalizedEvent, newPersonContainer] = await this.runStep(processPersonsStep, [
+                this,
+                processedEvent,
+                personContainer,
+            ])
+
+            const preparedEvent = await this.runStep(prepareEventStep, [this, normalizedEvent])
+
+            await this.runStep(createEventStep, [this, preparedEvent, newPersonContainer])
+            return this.registerLastStep('createEventStep', event.team_id, [preparedEvent, newPersonContainer])
+        } else {
+            return this.registerLastStep('pluginsProcessEventStep', event.team_id, [event])
+        }
```

Comment:
nit: could be a little bit easier to read in this order here and other places where we stop the flow
```suggestion
        if (processedEvent == null) {
            return this.registerLastStep('pluginsProcessEventStep', event.team_id, [event])
        } 
        const [normalizedEvent, newPersonContainer] = await this.runStep(processPersonsStep, [
                this,
                processedEvent,
                personContainer,
        ])

        const preparedEvent = await this.runStep(prepareEventStep, [this, normalizedEvent])

        await this.runStep(createEventStep, [this, preparedEvent, newPersonContainer])
        return this.registerLastStep('createEventStep', event.team_id, [preparedEvent, newPersonContainer])
```